Refactor Device to not depend on Backend.

ezyang opened the pull request:
```
- Removed Backend constructor from Device, and fixed all
  use-sites to use DeviceType::CPU instead of kCPU, or
  use a new function backendToDeviceType to perform
  the conversion.
- New method device_type() on Type; it gives you the
  underlying device type, e.g., CPU for SparseCPU.
- We add backward compatibility for kCPU/kCUDA uses,
  by introducing a new special type which is implicitly
  convertible to both DeviceType and Backend.  As long as
  you don't define a function that's overloaded on both
  DeviceType and Backend (but not on BackendOrDeviceType),
  the implicit conversions will ensure that uses
  of at::Device(at::kCPU) keep working. We fixed use-sites in
  the library, but did NOT fix sites in the test code, so that
  we can exercise this BC code.
```

Differential Revision: D9301861

goldsborough commented on aten/src/ATen/Device.h:
```diff
@@ -60,10 +46,8 @@ struct Device {
-  /* implicit */ Device(Backend backend, int32_t index = -1)
-      : Device(backend_to_type(backend), index) {}
+  /* implicit */ Device(BackendOrDeviceType backend_or_device_type)
+    : Device(backend_or_device_type.operator DeviceType()) {}
```
`backend_or_device_type.operator DeviceType()` 😮

ezyang replied:
I guess a `static_cast` would have been sufficient here. In another version of the patch, calling the operator was mandatory, because `static_cast` would also consider the constructors (which caused ambiguity).

goldsborough reviewed (approved):
Looks like a reasonable codemod! Some stylistic nits

ezyang commented on aten/src/ATen/Backend.h:
```diff
@@ -78,6 +84,42 @@ static inline TensorTypeId backendToTensorTypeId(Backend b) {
+// which only CPU is valid).  We've tightened up some APIs, but to
+// help user code out, define a class which supports implicit conversions
+// to both.
+class BackendOrDeviceType {
```
sure why not

ezyang commented on aten/src/ATen/Backend.h:
```diff
@@ -78,6 +84,42 @@ static inline TensorTypeId backendToTensorTypeId(Backend b) {
   }
 }
 
+static inline DeviceType backendToDeviceType(Backend b) {
```
Normally I'd put this on the class, but they're both enum classes so there's no where to put it.

gchanan replied:
This implementation is not what I was expecting from the function name (I was expecting the toDense version).

The reason being that "to" implies a map and there's a perfectly valid map for all non-undefined backends to device types.  I'd suggest either changing the implementation or renaming to something like "backendAsDeviceType", to emphasize this is the device type when the backend is used as a device type.

ezyang replied:
I think I am going to "densify" it.

ezyang commented on aten/src/ATen/core/typeid.cpp:
```diff
@@ -9,10 +9,26 @@
+  const at::DataType::underlying_type new_value = ++counter;
+  if (new_value ==
+      std::numeric_limits<at::DataType::underlying_type>::max()) {
+    throw std::logic_error(
```
I think this might be a bit more expensive in terms of binary size. Experiment coming.

gchanan reviewed (commented):
Do we really want/need this backwards compatibility stuff?  We don't guarantee backwards compatibility with the C++ API and kCUDA, kCPU are so convenient that people might continue to use them as devices (bad!).  I'd almost prefer reclaiming those names for the device types, and moving the backend short names to something else.

Thoughts @goldsborough?

ezyang commented:
Yes, I am somewhat convinced we should make `kCPU` and `kCUDA` device types. I guess I'll have to try it and see what breaks. Maybe later today.

gchanan reviewed (approved):
looks good, although tests are failing?

ezyang committed:
Refactor Device to not depend on Backend. (#10478)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10478

- Removed Backend constructor from Device, and fixed all
  use-sites to use DeviceType::CPU instead of kCPU, or
  use a new function backendToDeviceType to perform
  the conversion.
- New method device_type() on Type; it gives you the
  underlying device type, e.g., CPU for SparseCPU.
- We add backward compatibility for kCPU/kCUDA uses,
  by introducing a new special type which is implicitly
  convertible to both DeviceType and Backend.  As long as
  you don't define a function that's overloaded on both
  DeviceType and Backend (but not on BackendOrDeviceType),
  the implicit conversions will ensure that uses
  of at::Device(at::kCPU) keep working. We fixed use-sites in
  the library, but did NOT fix sites in the test code, so that
  we can exercise this BC code.

Reviewed By: Yangqing

Differential Revision: D9301861

fbshipit-source-id: 937fe15e32c2beba9235b8b26b9532ba85594635
